File: 1956-maximum-element-after-decreasing-and-rearranging/maximum-element-after-decreasing-and-rearranging.py
class Solution(object):
    def maximumElementAfterDecrementingAndRearranging(self, arr):
        l = len(arr)
        counter = [0] * l

        for i in range(l):
            counter[min(arr[i] - 1, l - 1)] += 1

        ans = 1
        for i in range(1, l):
            ans = min(i + 1, ans + counter[i])

        return ans
        

File: 1956-maximum-element-after-decreasing-and-rearranging/maximum-element-after-decreasing-and-rearranging.cpp
class Solution {
public:
    int maximumElementAfterDecrementingAndRearranging(vector<int>& arr) {
        int l = arr.size();
        std::vector<int> counter(l, 0);

        for (int i = 0; i < l; ++i) {
            ++counter[std::min(arr[i] - 1, l - 1)];
        }

        int ans = 1;
        for (int i = 1; i < l; ++i) {
            ans = std::min(i + 1, ans + counter[i]);
        }

        return ans;
    }
};